lib/reporting.rb

```diff
@@ -0,0 +1,163 @@
+  #
+  def export_users
+    export('user_status', users.keys, [users.values])
+  end
+
+  # | id | not_started | started | in_progress | completed | module_start | module_complete | confidence_check_start | confidence_check_complete | start_assessment | pass_assessment | fail_assessment |
+  # |----|-------------|---------|-------------|-----------|--------------|-----------------|------------------------|---------------------------|------------------|-----------------|-----------------|
+  # | 1  | 992         | 2071    | 862         | 1209      | 2071         | 1209            | 776                    | 832                       | 799              | 623             | 171             |
+  # | 2  | 2175        | 888     | 194         | 694       | 888          | 694             | 465                    | 480                       | 477              | 232             | 246             |
+  # | 3  | 2525        | 538     | 82          | 456       | 538          | 456             | 316                    | 332                       | 318              | 108             | 211             |
+  # | 4  | 2773        | 290     | 49          | 241       | 290          | 241             | 240                    | 240                       | 242              | 200             | 41              |
+  #
+  def export_modules
+    export('module_status', modules.first.keys, modules.map(&:values))
+  end
+
+  # ----------------------------------------------------------------------------
+
+  def users
+    {
+      # registration scopes
+      registration_complete: User.registration_complete.count,
+      registration_incomplete: User.registration_incomplete.count,
+      reregistered: User.reregistered.count,
+      registered_since_private_beta: User.registered_since_private_beta.count,
+      private_beta_only_registration_incomplete: User.private_beta_only_registration_incomplete.count,
+      private_beta_only_registration_complete: User.private_beta_only_registration_complete.count,
+
+      # registration events
+      registration_events: Ahoy::Event.user_registration.count,
+      private_beta_registration_events: Ahoy::Event.private_beta_registration.count,
+      public_beta_registration_events: Ahoy::Event.public_beta_registration.count,
+
+      # all
+      total: User.all.count,
+
+      # devise
+      locked_out: User.locked_out.count,
+      confirmed: User.confirmed.count,
+      unconfirmed: User.unconfirmed.count,
+
+      # user input
+      user_defined_roles: User.all.collect(&:role_type_other).uniq.count,
+
+      # course engagement
+      started_learning: started_learning,
+      not_started_learning: not_started_learning,
+
+      # Number of distinct users who have at least one note
+      with_notes: get_users_with_notes_count,
```

Comment:
Can I suggest that you make `get_users_with_notes_count` a private method and instead we can make assertions on the public `users[:with_notes]` and `users[:with_notes_percentage]` values.

Reply:
Good point, have updated [5ecd562](https://github.com/DFE-Digital/early-years-foundation-recovery/pull/622/commits/5ecd56260e922896853858f36986bf12cc6e1f08)